File: src/stages/git-analyzer.ts
import type { ChangedFile, CommitInfo, FileCategory, GitAnalysis } from '../types.js'
import { getLanguageFromPath, git, gitLines, truncateDiff } from '../utils.js'

const DEFAULT_MAX_DIFF_LINES_PER_FILE = 200
const DEFAULT_MAX_TOTAL_DIFF_LINES = 3000

function categorizeFile(path: string): FileCategory {
  const lower = path.toLowerCase()

  if (/\.(test|spec)\.[^.]+$/.test(lower) || /__tests__\//.test(lower) || /\/test\//.test(lower))
    return 'test'
  if (/\.(config|rc)\.[^.]+$/.test(lower) || /\/\.[^/]*rc/.test(lower) || /tsconfig/.test(lower))
    return 'config'
  if (/\.(css|scss|less|sass|styl)$/.test(lower))
    return 'style'
  if (/migrations?\//.test(lower) || /\.migration\.[^.]+$/.test(lower))
    return 'migration'
  if (/openapi|swagger|\.graphql$/.test(lower))
    return 'api-schema'
  if (/^\.github\/|jenkinsfile|dockerfile|\.gitlab-ci|\.circleci/i.test(lower))
    return 'ci'
  if (/\.md$/.test(lower) || /\/docs\//.test(lower))
    return 'docs'
  if (/\.(ts|tsx|js|jsx|py|go|rs|java|kt|rb|php|cs|swift|vue|svelte)$/.test(lower))
    return 'source'
  return 'other'
}

function parseStatus(letter: string): ChangedFile['status'] {
  switch (letter) {
    case 'A': return 'added'
    case 'D': return 'deleted'
    case 'R': return 'renamed'
    default: return 'modified'
  }
}

export async function analyzeGit(
  cwd: string,
  baseBranch: string,
  headBranch: string,
  options?: { maxDiffLinesPerFile?: number; maxDiffLines?: number },
): Promise<GitAnalysis> {
  const MAX_DIFF_LINES_PER_FILE = options?.maxDiffLinesPerFile ?? DEFAULT_MAX_DIFF_LINES_PER_FILE
  const MAX_TOTAL_DIFF_LINES = options?.maxDiffLines ?? DEFAULT_MAX_TOTAL_DIFF_LINES
  // Get merge base for accurate diff
  let mergeBase: string
  try {
    mergeBase = git(['merge-base', baseBranch, headBranch], cwd)
  } catch {
    mergeBase = baseBranch
  }

  // Get changed files with stats
  const nameStatusLines = gitLines(['diff', '--name-status', `${mergeBase}...${headBranch}`], cwd)

  const numstatLines = gitLines(['diff', '--numstat', `${mergeBase}...${headBranch}`], cwd)

  // Build numstat map
  const numstatMap = new Map<string, { additions: number; deletions: number }>()
  for (const line of numstatLines) {
    const [add, del, ...pathParts] = line.split('\t')
    const filePath = pathParts.join('\t') // handle renames with =>
    numstatMap.set(filePath, {
      additions: add === '-' ? 0 : parseInt(add, 10),
      deletions: del === '-' ? 0 : parseInt(del, 10),
    })
  }

  // Build changed files
  const changedFiles: ChangedFile[] = []
  for (const line of nameStatusLines) {
    const parts = line.split('\t')
    const statusLetter = parts[0][0]
    const filePath = statusLetter === 'R' ? parts[2] : parts[1]
    if (!filePath) continue

    const stats = numstatMap.get(filePath) ??
      numstatMap.get(parts.slice(1).join('\t')) ??
      { additions: 0, deletions: 0 }

    let diff = ''
    try {
      diff = git(['diff', `${mergeBase}...${headBranch}`, '--', filePath], cwd)
      diff = truncateDiff(diff, MAX_DIFF_LINES_PER_FILE)
    } catch {
      diff = '(unable to retrieve diff)'
    }

    changedFiles.push({
      path: filePath,
      status: parseStatus(statusLetter),
      additions: stats.additions,
      deletions: stats.deletions,
      diff,
      language: getLanguageFromPath(filePath),
      category: categorizeFile(filePath),
    })
  }

  // Truncate total diff if needed
  let totalLines = 0
  for (const f of changedFiles) {
    const lineCount = f.diff.split('\n').length
    totalLines += lineCount
    if (totalLines > MAX_TOTAL_DIFF_LINES) {
      f.diff = truncateDiff(f.diff, Math.max(20, MAX_DIFF_LINES_PER_FILE - (totalLines - MAX_TOTAL_DIFF_LINES)))
    }
  }

  // Get commits
  const commitLines = gitLines(['log', '--format=%H|%s|%ai|%an', `${mergeBase}...${headBranch}`], cwd)
  const commits: CommitInfo[] = commitLines.map(line => {
    const [hash, message, date, author] = line.split('|')
    return { hash, message, date, author }
  })

  // Aggregate stats
  const stats = changedFiles.reduce(
    (acc, f) => ({
      additions: acc.additions + f.additions,
      deletions: acc.deletions + f.deletions,
      filesChanged: acc.filesChanged + 1,
    }),
    { additions: 0, deletions: 0, filesChanged: 0 },
  )

  return { baseBranch, headBranch, changedFiles, stats, commits }
}

File: rust/src/stages/git_analyzer.rs
use anyhow::Result;
use regex::Regex;

use crate::types::{ChangedFile, CommitInfo, FileCategory, FileStatus, GitAnalysis, GitStats};
use crate::utils::{get_language_from_path, git, git_lines, truncate_diff};

const DEFAULT_MAX_DIFF_LINES_PER_FILE: usize = 200;
const DEFAULT_MAX_TOTAL_DIFF_LINES: usize = 3000;

fn categorize_file(path: &str) -> FileCategory {
    let lower = path.to_lowercase();

    if Regex::new(r"\.(test|spec)\.[^.]+$")
        .unwrap()
        .is_match(&lower)
        || lower.contains("__tests__/")
        || lower.contains("/test/")
    {
        return FileCategory::Test;
    }
    if Regex::new(r"\.(config|rc)\.[^.]+$")
        .unwrap()
        .is_match(&lower)
        || Regex::new(r"/\.[^/]*rc").unwrap().is_match(&lower)
        || lower.contains("tsconfig")
    {
        return FileCategory::Config;
    }
    if Regex::new(r"\.(css|scss|less|sass|styl)$")
        .unwrap()
        .is_match(&lower)
    {
        return FileCategory::Style;
    }
    if lower.contains("migration") || Regex::new(r"\.migration\.[^.]+$").unwrap().is_match(&lower) {
        return FileCategory::Migration;
    }
    if lower.contains("openapi") || lower.contains("swagger") || lower.ends_with(".graphql") {
        return FileCategory::ApiSchema;
    }
    if lower.starts_with(".github/")
        || lower.contains("jenkinsfile")
        || lower.contains("dockerfile")
        || lower.contains(".gitlab-ci")
        || lower.contains(".circleci")
    {
        return FileCategory::Ci;
    }
    if lower.ends_with(".md") || lower.contains("/docs/") {
        return FileCategory::Docs;
    }
    if Regex::new(r"\.(ts|tsx|js|jsx|py|go|rs|java|kt|rb|php|cs|swift|vue|svelte)$")
        .unwrap()
        .is_match(&lower)
    {
        return FileCategory::Source;
    }
    FileCategory::Other
}

fn parse_status(letter: char) -> FileStatus {
    match letter {
        'A' => FileStatus::Added,
        'D' => FileStatus::Deleted,
        'R' => FileStatus::Renamed,
        _ => FileStatus::Modified,
    }
}

pub async fn analyze_git(
    cwd: &str,
    base_branch: &str,
    head_branch: &str,
    max_diff_lines_per_file: Option<usize>,
    max_diff_lines: Option<usize>,
) -> Result<GitAnalysis> {
    let max_per_file = max_diff_lines_per_file.unwrap_or(DEFAULT_MAX_DIFF_LINES_PER_FILE);
    let max_total = max_diff_lines.unwrap_or(DEFAULT_MAX_TOTAL_DIFF_LINES);

    // Get merge base
    let merge_base = git(&["merge-base", base_branch, head_branch], cwd)
        .unwrap_or_else(|_| base_branch.to_string());

    let range = format!("{}...{}", merge_base, head_branch);

    // Get changed files with name-status
    let name_status_lines = git_lines(&["diff", "--name-status", &range], cwd).unwrap_or_default();

    // Get numstat
    let numstat_lines = git_lines(&["diff", "--numstat", &range], cwd).unwrap_or_default();

    // Build numstat map
    let mut numstat_map = std::collections::HashMap::new();
    for line in &numstat_lines {
        let parts: Vec<&str> = line.splitn(3, '\t').collect();
        if parts.len() >= 3 {
            let additions = if parts[0] == "-" {
                0
            } else {
                parts[0].parse::<usize>().unwrap_or(0)
            };
            let deletions = if parts[1] == "-" {
                0
            } else {
                parts[1].parse::<usize>().unwrap_or(0)
            };
            numstat_map.insert(parts[2].to_string(), (additions, deletions));
        }
    }

    // Build changed files
    let mut changed_files: Vec<ChangedFile> = Vec::new();
    for line in &name_status_lines {
        let parts: Vec<&str> = line.split('\t').collect();
        if parts.is_empty() {
            continue;
        }
        let status_letter = parts[0].chars().next().unwrap_or('M');
        let file_path = if status_letter == 'R' {
            parts.get(2).unwrap_or(&"")
        } else {
            parts.get(1).unwrap_or(&"")
        };
        if file_path.is_empty() {
            continue;
        }

        let (additions, deletions) = numstat_map
            .get(*file_path)
            .or_else(|| {
                let joined = parts[1..].join("\t");
                numstat_map.get(&joined)
            })
            .copied()
            .unwrap_or((0, 0));

        let diff = match git(&["diff", &range, "--", file_path], cwd) {
            Ok(d) => truncate_diff(&d, max_per_file),
            Err(_) => "(unable to retrieve diff)".to_string(),
        };

        changed_files.push(ChangedFile {
            path: file_path.to_string(),
            status: parse_status(status_letter),
            additions,
            deletions,
            diff,
            language: get_language_from_path(file_path),
            category: categorize_file(file_path),
        });
    }

    // Truncate total diff if needed
    let mut total_lines = 0;
    for f in &mut changed_files {
        let line_count = f.diff.lines().count();
        total_lines += line_count;
        if total_lines > max_total {
            let new_max = max_per_file.saturating_sub(total_lines - max_total).max(20);
            f.diff = truncate_diff(&f.diff, new_max);
        }
    }

    // Get commits
    let commit_lines =
        git_lines(&["log", "--format=%H|%s|%ai|%an", &range], cwd).unwrap_or_default();

    let commits: Vec<CommitInfo> = commit_lines
        .iter()
        .map(|line| {
            let parts: Vec<&str> = line.splitn(4, '|').collect();
            CommitInfo {
                hash: parts.first().unwrap_or(&"").to_string(),
                message: parts.get(1).unwrap_or(&"").to_string(),
                date: parts.get(2).unwrap_or(&"").to_string(),
                author: parts.get(3).unwrap_or(&"").to_string(),
            }
        })
        .collect();

    // Aggregate stats
    let stats = changed_files.iter().fold(
        GitStats {
            additions: 0,
            deletions: 0,
            files_changed: 0,
        },
        |acc, f| GitStats {
            additions: acc.additions + f.additions,
            deletions: acc.deletions + f.deletions,
            files_changed: acc.files_changed + 1,
        },
    );

    Ok(GitAnalysis {
        base_branch: base_branch.to_string(),
        head_branch: head_branch.to_string(),
        changed_files,
        stats,
        commits,
    })
}
